src/driver.rs

```diff
@@ -0,0 +1,299 @@
+  /// a larger number then finish we will execute all the "sub" motions until we
+  /// reach finish, and if the start number is smaller we will execute all the
+  /// "add" motions until we reach finish.
+  fn execute(&mut self, motions: &Vec<Motion>, some_start: isize, some_finish: isize) -> Result<(), Self::E> {
+    let start = clamp(some_start, 0, motions.len() as isize);
+    let finish = clamp(some_finish, 0, motions.len() as isize);
+    if start != finish {
+      let operation = Operation::get(finish, start);
+      let mut i = start;
+      loop {
+        if i == finish {
+          break;
+        }
+
+        if operation == Operation::Sub {
+          i += Operation::sub();
+        }
+
+        match self.execute_motion(&motions[i as usize].add) {
+          Ok(_) => (),
+          // Handle execution error by first trying to set the new status.
+          Err(err) => {
+            try!(self.set_status(i));
+            return Err(err);
+          }
+        }
+
+        if operation == Operation::Add {
+          i += Operation::add();
+        }
+      }
+      self.set_status(i)
+    } else {
+      Ok(())
+    }
+  }
+
+  /// Starts at the current driver status and executes n motions in either the
+  /// "add" or "sub" direction.
+  fn shift(&mut self, motions: &Vec<Motion>, n: isize) -> Result<(), Self::E> {
+    let start = try!(self.get_status());
+    let finish = clamp(start + n, 0, motions.len() as isize);
+    self.execute(motions, start, finish)
+  }
+
+
+  /// Executes as many motions as necessary to move the driver from its current
+  /// status to a certain state.
+  fn goto(&mut self, motions: &Vec<Motion>, finish: isize) -> Result<(), Self::E> {
+    let status = try!(self.get_status());
+    self.execute(motions, status, finish)
+  }
+
+  /// "sub's" the last motion then "add's" it back.
+  fn redo(&mut self, motions: &Vec<Motion>) -> Result<(), Self::E> {
+    try!(self.shift(motions, -1));
+    self.shift(motions, 1)
+  }
+
+  /// Applies all remaining "add" motions to the driver.
+  fn up(&mut self, motions: &Vec<Motion>) -> Result<(), Self::E> {
+    let last = motions.len() as isize;
+    self.goto(motions, last)
+  }
+
+  /// Applies all remaining "sub" motions to the driver.
+  fn down(&mut self, motions: &Vec<Motion>) -> Result<(), Self::E> {
+    self.goto(motions, 0)
+  }
+
+  /// Applies all remaining "sub" motions to the driver before adding them all
+  /// the subbed motions back.
+  fn reset(&mut self, motions: &Vec<Motion>) -> Result<(), Self::E> {
+    let status = try!(self.get_status());
+    try!(self.execute(motions, status, 0));
+    self.execute(motions, 0, status)
+  }
+}
+
+/// Forces an integer to be inside a range. Similar to using both the algebraic
+/// `min` and `max` functions with the specified values.
+fn clamp(n: isize, min: isize, max: isize) -> isize {
+  if n < min {
+    return min;
+  }
+  if n > max {
+    return max;
+  }
+  n
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
The unit tests? I moved them because I'm pretty sure it's idiomatic rust. I've seen it when I've gone source diving a couple of times. In addition [cargo](http://doc.crates.io/guide.html#tests) and [the rust book](https://doc.rust-lang.org/book/testing.html) seem to recommend this method.